Lib/zipfile.py

```diff
@@ -1689,6 +1713,52 @@ def _extract_member(self, member, targetpath, pwd):
             shutil.copyfileobj(source, target)
 
         return targetpath
+    
+    def _remove_member(self, member):
+        # get a sorted filelist by header offset, in case the dir order
+        # doesn't match the actual entry order
+        fp = self.fp
+        entry_offset = 0
+        filelist = sorted(self.filelist, key=attrgetter('header_offset'))
+        for i in range(len(filelist)):
+            info = filelist[i]
+            # find the target member
+            if info.header_offset < member.header_offset:
+                continue
+
+            # get the total size of the entry
+            entry_size = None
+            if i == len(filelist) - 1:
+                entry_size = self.start_dir - info.header_offset
+            else:
+                entry_size = filelist[i + 1].header_offset - info.header_offset
```

Comment:
Is there anything in zipfile spec which says the header offsets have to be contiguous?

Reply:
It doesn't matter anyway. The implementation moves all bytes between the start of a section and the start of the next section. If there were to be random bytes between the end of a section and the start of the next section, they will be moved too.